x-pack/legacy/plugins/file_upload/public/util/file_parser.js

```diff
@@ -7,39 +7,121 @@
           defaultMessage: 'Error, no file provided',
-        })));
-    }
-    const fr = new window.FileReader();
-    fr.onload = e => resolve(e.target.result);
-    fr.onerror = () => {
-      fr.abort();
+        })
+      )
+    );
+  }
+
+
+  // Halt any previous file reading & pattern checking activity
+  if (prevFileReader) {
+    prevFileReader.abort();
+  }
+  if (prevPatternReader) {
+    prevPatternReader.abortStream();
+  }
+
+  // Set up feature tracking
+  let featuresProcessed = 0;
+  const onFeatureRead = feature => {
+    // TODO: Add handling and tracking for cleanAndValidate fails
+    featuresProcessed++;
+    return cleanAndValidate(feature);
+  };
+
+  let start;
+  let stop = fileBuffer;
+
+  prevFileReader = fileReader;
+
+  const patternReader = new PatternReader();
+  patternReader.onGeoJSONFeaturePatternDetect(onFeatureRead);
+  prevPatternReader = patternReader;
+
+  const filePromise = new Promise((resolve, reject) => {
+    fileReader.onloadend = ({ target: { readyState, result } }) => {
+      if (readyState === FileReader.DONE) {
+        if (!getFileParseActive() || !result) {
+          fileReader.abort();
+          patternReader.abortStream();
+          resolve(null);
+          return;
+        }
+        chunkHandler({
+          featuresProcessed,
+          bytesProcessed: stop || file.size,
+          totalBytes: file.size
+        });
+        patternReader.writeDataToPatternStream(result);
+        if (!stop) {
+          return;
+        }
+
+        start = stop;
+        const newStop = stop + fileBuffer;
+        // Check EOF
+        stop = newStop > file.size ? undefined : newStop;
+        readSlice(fileReader, file, start, stop);
+      }
+    };
+    fileReader.onerror = () => {
+      fileReader.abort();
+      patternReader.abortStream();
       reject(new Error(i18n.translate(
         'xpack.fileUpload.fileParser.errorReadingFile', {
           defaultMessage: 'Error reading file',
         })));
     };
-    fr.readAsText(file);
+    patternReader.onStreamComplete(parsedGeojson => {
+      if (!featuresProcessed) {
+        reject(
+          new Error(
+            i18n.translate('xpack.fileUpload.fileParser.noFeaturesDetected', {
+              defaultMessage: 'Error, no features detected',
+            })
+          )
+        );
+      } else {
+        resolve(parsedGeojson);
+      }
+    });
   });
-
-  return await readPromise;
-}
+  readSlice(fileReader, file, start, stop);
+  return filePromise;
+};
 
 export function jsonPreview(json, previewFunction) {
-  // Call preview (if any)
   if (json && previewFunction) {
     const defaultName = _.get(json, 'name', 'Import File');
     previewFunction(json, defaultName);
   }
 }
 
-export async function parseFile(file, transformDetails, previewCallback = null) {
+export async function parseFile(
+  file, transformDetails, previewCallback = null, onChunkParse, getFileParseActive
```

Comment:
use named variables.

Consider initializing `previewCallback = () => {}`, then you don't need to do a check on `previewFunction` existnece https://github.com/elastic/kibana/pull/46710/files#diff-b3b663abe7760e32b90cbccdd89d8e9cR116

Reply:
Done! Leaving `previewCallback` as is for now since it wasn't affected to heavily by this PR and is about to effectively "go away" in its current form with the consolidation of preview and indexing chunking logic

Reply:
There's no need to leave crufty code in the app for possible future changes that have not merged yet. You may as well remove it here since it is flagged in this PR. Otherwise, link it to an existing issue explaining the tech debt.